### ISSUE TITLE
configfailurepolicy=continue only works for BeforeTest when using TestNG XML file

### ISSUE THREAD

**Issue by Mei-self** [testng-team/testng#2731]
### TestNG Version
> 7.5

### Expected behavior
if configfailurepolicy=continue and any BeforeXXX method fails, the remaining test method should execute not skip

### Actual behavior
only BeforeTest fails, the remaining test method would execute

### Is the issue reproducible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### Test case sample
>  
```
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="TestSuite" configfailurepolicy="continue">
    <test name="Test">
        <classes>
            <class name="TestConfigFail"/>
        </classes>
    </test>
</suite>
```

```
public class TestConfigFail {

    @BeforeSuite
    public void beforeSuite() {
        Assert.fail("This before suite is fail");
    }

    @BeforeTest
    public void beforeTest() {
        Assert.fail("This before test is fail");
    }

    @BeforeClass
    public void beforeClass(){
        Assert.fail("This before class is fail");
    }

    @BeforeMethod
    public void beforeMethod(){
        Assert.fail("This before method is fail");
    }

    @Test
    public void test() {
        System.out.println("This is test method");
    }
}

```
### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

**Comment by krmahadevan** [comment]
@Wang-Meimei - Is this something that was broken in `7.5` or has this never worked ? Can u please check in `7.4.0` and `6.14.3` and post back your findings ?

**Comment by Wang-Meimei** [comment]
> @Wang-Meimei - Is this something that was broken in `7.5` or has this never worked ? Can u please check in `7.4.0` and `6.14.3` and post back your findings ?

It seems this never worked. But from the document, it should work. So maybe need to update document?

**Comment by juherr** [comment]
> It seems this never worked. But from the document, it should work. So maybe need to update document?

Yes, as we know now it is no a regression, we need to determine the expected behavior and fix the code or the doc.

I think the issue is valid and the method should be run.
@krmahadevan wdyt?

**Comment by krmahadevan** [comment]
> > It seems this never worked. But from the document, it should work. So maybe need to update document?
> 
> Yes, as we know now it is no a regression, we need to determine the expected behavior and fix the code or the doc.
> 
> I think the issue is valid and the method should be run. @krmahadevan wdyt?

I dont know if this is an issue @juherr . 

If there is a higher and lower order configuration, then we should not try to execute the lower order configuration.

For .e.g,

Lets say I have

* `@BeforeSuite`
* `@BeforeTest`
* `@Test`

In this case, we should not try to execute the `@BeforeTest` because this is a configuration method. But then one can argue that if one has to get to the level of running a `@Test` method, then all the above level configurations need to be run.

So am not sure what the behaviour should be. 

@bj-9527 - Please note.

**Comment by juherr** [comment]
IMO, the only difference between `continue` and `skip` is that `continue` will execute the method, and `skip` won't execute and will directly set the SKIP status.

But when I see #2735 changes in tests, I wonder if the behavior change should not be configurable.

**Comment by bj-9527** [comment]
It could be config by `FailurePolicy` which could be set as `SKIP` or `CONTINUE`, then the behavior could change.

> So am not sure what the behaviour should be.

For, this change, the `SKIP` are working as before, the `CONTINUE` will execute the following method even previous config failed.

**Comment by juherr** [comment]
*(Written by Claude, on Julien's behalf.)*

Coming back to this after three years, with the behaviour measured rather than reasoned about. The title of this issue is literally accurate. On current `master`, one failing `@Before*` method plus one `@Test`, run under each policy:

| Failing method | `skip` | `continue` |
|---|---|---|
| `@BeforeSuite` | test skipped | test skipped |
| `@BeforeTest` | test skipped | **test runs** |
| `@BeforeClass` | test skipped | test skipped |
| `@BeforeMethod` | test skipped | test skipped |

The configuration failure is reported in all eight cases, so `continue` never turns a broken setup into a green run — it only decides whether the test method is attempted.

### Why the four levels differ

They are not four variations of one rule; the failure is recorded in three different places (`ConfigInvoker.recordConfigurationInvocationFailed`):

- **`@BeforeSuite`** sets `m_suiteState.failed()`, and `hasConfigurationFailureFor` short-circuits on that *before* consulting the policy at all. Consistent with `ignoreFailure` (#2862) having deliberately been added to `@BeforeTest`/`@BeforeClass`/`@BeforeMethod` but **not** to `@BeforeSuite` — a suite-level failure is meant to be an unconditional stop.
- **`@BeforeClass`** and **`@BeforeMethod`** are recorded against the failing instance / the failing invocation token. `continue` does honour the policy here, but only by *narrowing* what the failure invalidates instead of poisoning the whole class. With a single instance and a single invocation there is nothing left to narrow, so the test still skips. That narrowing is what #990, #2148 and #2862 are built on, and what every `CONTINUE` expectation in `FailurePolicyTest` encodes.
- **`@BeforeTest`** is recorded against *no* instance (it is invoked without one, so the sentinel `NULL_OBJECT` is stored). Under `continue` the check is `set.contains(instance)`, which can never match a real instance — so a failed `@BeforeTest` invalidates nothing and the tests run.

So @Mei-self's original reading of the docs was right *for one level out of four*, and @krmahadevan's objection was right for the other three.

### The documentation says both things

- The shipped DTD and `TestNG.setConfigFailurePolicy`: *"whether to continue **attempting** Before/After Class/Methods **after they've failed once**"* — the narrowing reading, matching `@BeforeClass`/`@BeforeMethod`.
- The website: *"continue to execute the **remaining tests** in the suite"* — the permissive reading, matching only `@BeforeTest`.
- `XmlSuite.FailurePolicy` itself: nothing at all.

### What has been done, and what has not

Two PRs, neither of which changes any behaviour:

- testng-team/testng#3426 — a characterization test for the whole 4×2 matrix (`FailurePolicyTest` covered neither `@BeforeSuite` nor `@BeforeTest`, under either policy, which is the gap this issue went through), plus javadoc on `XmlSuite.FailurePolicy` and the accessors that matches the matrix.
- testng-team/testng-team.github.io#85 — the website row, rewritten around the DTD's framing.

**This issue stays open**, because the real question is still the one @krmahadevan asked in 2022 and it is a design call, not a documentation one:

> is `@BeforeTest` + `continue` the **bug** — too permissive, and it should skip like `@BeforeClass` does — or is it the **model**, and the other levels are too strict?

Both answers break somebody, so it needs a maintainer decision rather than an inference from the docs. What #3426 buys is that the decision is now cheap: whichever way it goes, the expectations that have to move are named in one place.

@krmahadevan @juherr — over to you on that one.

**Comment by juherr** [comment]
**Verdict: still relevant**

Current behavior is now explicitly pinned by merged PR #3426 and its `GITHUB-2731` characterization test: with `configfailurepolicy=continue`, a failed `@BeforeTest` lets the test method run, while failed `@BeforeSuite`, `@BeforeClass`, and `@BeforeMethod` still leave the single affected test skipped. #3426 deliberately changed only tests/documentation, not those semantics.

So the original inconsistency is confirmed, but the fix direction remains a compatibility/design decision: make `@BeforeTest` stricter, or make the other levels more permissive. The existing characterization matrix provides the exact regression surface once that decision is made.

I did not execute the characterization test or the historical reproducer during this triage.
